fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java

```diff
@@ -14,76 +14,286 @@
+
+                Expr variableResult = queryStatement.getQueryRelation().getOutputExpression().get(0);
+
+                //can not apply to numeric types or complex type are not supported
+                if (variableResult.getType().isOnlyMetricType() || variableResult.getType().isFunctionType()
+                        || variableResult.getType().isComplexType()) {
+                    throw new SemanticException("Can't set variable with type " + variableResult.getType());
+                }
+
+                ((SelectRelation) queryStatement.getQueryRelation()).getSelectList().getItems().set(0,
+                        new SelectListItem(new CastExpr(Type.VARCHAR, variableResult), null));
+
+                Subquery subquery = new Subquery(queryStatement);
+                subquery.setType(variableResult.getType());
+                var.setUnevaluatedExpression(subquery);
             }
         }
     }
+
+    private static void analyzeSetUserPropertyVar(SetUserPropertyVar setUserPropertyVar) {
+        if (Strings.isNullOrEmpty(setUserPropertyVar.getPropertyKey())) {
+            throw new SemanticException("User property key is null");
+        }
+
+        if (setUserPropertyVar.getPropertyValue() == null) {
+            throw new SemanticException("User property value is null");
+        }
+
+        if (!setUserPropertyVar.getPropertyKey().equals("max_user_connections")) {
+            throw new SemanticException("Unknown property key: " + setUserPropertyVar.getPropertyKey());
+        }
+    }
+
+    private static void analyzeSetNames(SetNamesVar var) {
+        String charset = var.getCharset();
+
+        if (Strings.isNullOrEmpty(charset)) {
+            charset = SetNamesVar.DEFAULT_NAMES;
+        } else {
+            charset = charset.toLowerCase();
+        }
+        // utf8-superset transform to utf8
+        if (charset.startsWith(SetNamesVar.DEFAULT_NAMES)) {
+            charset = SetNamesVar.DEFAULT_NAMES;
+        }
+
+        if (!charset.equalsIgnoreCase(SetNamesVar.DEFAULT_NAMES) && !charset.equalsIgnoreCase(SetNamesVar.GBK_NAMES)) {
+            ErrorReport.reportSemanticException(ErrorCode.ERR_UNKNOWN_CHARACTER_SET, charset);
+        }
+        // be is not supported yet,so Display unsupported information to the user
+        if (!charset.equalsIgnoreCase(SetNamesVar.DEFAULT_NAMES)) {
+            throw new SemanticException("charset name " + charset + " is not supported yet");
+        }
+
+        var.setCharset(charset);
+    }
+
+    private static void analyzeSetPassVar(SetPassVar var, ConnectContext session) {
+        try {
+            UserIdentity userIdentity = var.getUserIdent();
+            if (userIdentity == null) {
+                userIdentity = session.getCurrentUserIdentity();
+            }
+            userIdentity.analyze();
+            var.setUserIdent(userIdentity);
+            var.setPasswdBytes(MysqlPassword.checkPassword(var.getPasswdParam()));
```

Comment:
I see some auth process in the old code(`SetPassVar::analyze`). It is no longer needed?

Reply:
Yes, it's no longer needed, the new version's permission logic has been moved into PrivilegeCheckerV2